doc/modules/ROOT/pages/appendix/gds-session.adoc

```diff
@@ -0,0 +1,310 @@
+
+
+== GDS Session management
+
+The `GdsSessions` object is the top-level API which allows you three operations:
+
+- `get_or_create`: Create a new GDS Session, or connect to an existing one.
+- `list`: List all currently active GDS Sessions.
+- `delete`: Delete a GDS Session.
+
+
+=== Pre-requisites
+
+To create a GDS Session, you need to have a Neo4j Aura account.
+Using your account, you need to create Aura API credentials.
+These credentials are a pair of `CLIENT_ID` and `CLIENT_SECRET` that you will use to authenticate with the Aura API.
+For how to create these credentials, see the https://neo4j.com/docs/aura/platform/api/authentication/#_creating_credentials[Aura documentation].
+
+Using these credentials, you will create a `GdsSessions` object, and use it to create a new GDS Session.
+
+.Creating a GdsSessions object:
+[source, python, role=no-test]
+----
+from graphdatascience.session import GdsSessions, AuraAPICredentials
+
+CLIENT_ID = "my-aura-api-client-id"
+CLIENT_SECRET = "my-aura-api-client-secret"
+
+# Create a new GdsSessions object
+sessions = GdsSessions(ds_connection=AuraAPICredentials(CLIENT_ID, CLIENT_SECRET))
+----
+
+
+=== Creating a GDS Session
+
+To create a GDS Session, you need to provide the following information:
+
+- **Session name**.
+The name must be unique.
+
+- **Session size**.
+The size determines the amount of memory and CPU available to the session.
+It also determines the cost of running the session.
+Available sizes are listed in xref:gds-session-sizes[].
+
+- **DBMS connection**.
+This is a `DbmsConnectionInfo` object that contains the URI of an AuraDB instance, a username, and a password.
+
+With these pieces of information, we are ready to create a GDS Session.
+
+.Creating a GDS Session:
+[source, python, role=no-test]
+----
+from graphdatascience.session import SessionSizes, DbmsConnectionInfo
+
+name = "my-new-session"
+size = SessionSizes.by_memory().DEFAULT
+db_connection_info = DbmsConnectionInfo("neo4j+s://mydbid.databases.neo4j.io", "my-user", "my-password")
+
+gds = sessions.get_or_create(
+    session_name=name,
+    size=size,
+    db_connection=db_connection_info,
+)
+----
+
+The `get_or_create()` method will create a new GDS Session if it does not exist, or connect to an existing one if it does.
+If the session size is different from the existing one, an error will be raised.
+
+The return value of `get_or_create()` is an `AuraGraphDataScience` object.
+It offers a similar API to the `GraphDataScience` object, but it is configured to run on a GDS Session.
+By convention, always use the variable name `gds` for the return value of `get_or_create()`.
+
+
+[[gds-session-sizes]]
+==== Session sizes
+
+Session sizes are logical and specified using the `SessionSizes.by_memory()` enum.
+Possible values are `DEFAULT`, `XS`, `S`, `SM`, `M`, `ML`, `L`, `XL`, `XXL`, `X3L`, `X4L`, `X5L`.
```

Comment:
My ambition here is to map the logical size to a monetary cost, but not necessarily to a physical size. So I'm endeavouring to not expose the actual memory used. Now that is perhaps a bit strict; we may as well have a mapping layer that goes from "physical size" to actual physical size.

But to your direct Qs:

- DEFAULT = M
- No

Reply:
I see. Yet, there is a mention of a `by_memory` enum, which would lead me to think they are still somewhat related to physical memory.

How can we explain a little better what they mean without being too specific? If we don't explain, a user would have no way to choose the appropriate one.

Reply:
Yeah there was a requirement for having two dimensions that sessions would be sized after: a CPU-scaling one and a Memory-scaling one. So far we only have implemented one of the size dimensions, but didn't want to occupy the namespace to block us in the future, hence `by_memory`. Without the other dimension it doesn't look good, I agree with that. I will note that down and see if we can come up with something better.

We've asked explicitly about feedback on the sizing API but received only a little. It is indeed related to physical memory, but my ambition is to not work with concrete numbers of bytes, rather preferring relative measures.

We are working on exposing a size estimator tool, which would recommend a particular size based on a user's provided node/rel counts and maybe selected algorithms (think AuraDS console; same underlying software). 

I think in principle, it is often difficult to know how large of a cloud instance you need upfront. But switching should be easy when you get it too small. I would argue that applies also here. Without the above tool, I could think of adding a part here recommending something like

>Start out with a smaller instance. If it is too small, delete it and create a larger one.

It's pretty raw. Do you think it is better than nothing?

With the tool available, I think of something like

>Use the tool to get an initial size recommendation. If it turns out to be too small, delete it and create a larger one.

The idea here is that since the GDS Session doesn't represent any persistent state, deleting and starting over is an OK thing to do. This mirrors my own experience of using e.g. AWS EC2.

Reply:
Ah, I see. Then I agree, let's wait for the estimator tool so we have something useful to say.